### ISSUE TITLE
Associate types problem

### ISSUE THREAD

**Issue by edwinb** [idris-lang/Idris-dev#2349]
Now that we've decided we should support associated types, I think the following should work:

    class ListLike a where
        Elem : Type
        fromList : List Elem -> a
  
    data Silly = None | Cons Int Silly
  
    instance ListLike Silly where
          Elem = Int
          fromList [] = None
          fromList (x :: xs) = Cons x (fromList xs)

But, alas, I get an error:

        Type mismatch between
                List Main.Silly instance of Main.ListLike, method Elem (Type of xs)
        and
                List Elem (Expected type)
        
        Specifically:
                Type mismatch between
                        Int
                and
                        Elem

I suspect this is a type class resolution problem in the body of the instance - it can be worked around by defining a helper function of type `List Int -> Silly`.

**Comment by david-christiansen** [comment]
I agree that this should work. I'll see if I can find time to poke it soon.

**Comment by edwinb** [comment]
This happens because when we come to resolve the type class in the definition of `fromList`, the instance isn't completely defined so while the instance is *resolved* it won't *reduce*, and so we get the mismatch between `Int` and `Elem`.

This may go beyond associated types and be a problem for some dependent type classes. The immediately obvious (but very fiddly) solution would be to add a partially defined typeclass between elaboration of each method.

**Comment by edwinb** [comment]
In fact, it's more general even that that (just noting thoughts here since I'm sitting in a conferece talk ;))

Instances are elaborated as a top level function plus where clauses. Effectively, one of the functions in the where block needs to use the reduction behaviour of the top level function *before* the where block is completely elaborated, but this is currently not supported at all.

So I think a proper solution involves changing the way `where` is elaborated to take this into account.

**Comment by david-christiansen** [comment]
It seems to me that it might work in this limited case to do a quick check on the methods, and when they don't pattern-match, elaborate them as let-bound lambdas instead. Is that crazy?

**Comment by ahmadsalim** [comment]
I think this would be really nice if we got it working.

**Comment by RyanGlScott** [comment]
The program in https://github.com/idris-lang/Idris-dev/issues/2349#issue-84960484 typechecks in Idris 1.3.0. Should we add a regression test for this?
